groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+
+                rc = ntco::IoRingUtil::enter(d_ring,
+                                             numToSubmit,
+                                             minimumToComplete);
+
+                NTCO_IORING_LOG_ENTER_COMPLETE(numToSubmit,
+                                               minimumToComplete,
+                                               rc);
+
+                if (rc < 0) {
+                    const int lastError = errno;
+                    if (lastError == ETIME) {
+                        NTCO_IORING_LOG_WAIT_TIMEOUT();
+                        break;
+                    }
+                    else {
+                        ntsa::Error error(lastError);
+                        NTCO_IORING_LOG_WAIT_FAILURE(error);
+                        break;
+                    }
+                }
+
+                continue;
+            }
+
+            if (entryCount == 1 && entryList[0].event() == 0) {
+                NTCO_IORING_LOG_WAIT_TIMEOUT();
+            }
+            else {
+                NTCO_IORING_LOG_WAIT_RESULT(entryCount);
+            }
+
+            break;
+        }
+
+        return entryCount;
+    }
+}
+
+bsl::size_t IoRingDevice::flush(ntco::IoRingCompletion* entryList,
+                                bsl::size_t             entryListCapacity)
+{
+    return d_completionQueue.pop(entryList, entryListCapacity);
+}
+
+// Return the index of the head entry in the submission queue.
+bsl::uint32_t IoRingDevice::submissionQueueHead() const
+{
+    return d_submissionQueue.head();
+}
+
+// Return the index of the tail entry in the submission queue.
+bsl::uint32_t IoRingDevice::submissionQueueTail() const
+{
+    return d_submissionQueue.tail();
+}
+
+// Return the maximum number of entries in the submission queue.
+bsl::uint32_t IoRingDevice::submissionQueueCapacity() const
+{
+    return d_submissionQueue.capacity();
+}
+
+// Return the index of the head entry in the completion queue.
+bsl::uint32_t IoRingDevice::completionQueueHead() const
+{
+    return d_completionQueue.head();
+}
+
+// Return the index of the tail entry in the completion queue.
+bsl::uint32_t IoRingDevice::completionQueueTail() const
+{
+    return d_completionQueue.tail();
+}
+
+// Return the maximum number of entries in the completion queue.
+bsl::uint32_t IoRingDevice::completionQueueCapacity() const
+{
+    return d_completionQueue.capacity();
+}
+
+bool IoRingDevice::supportsOperation(
+    ntco::IoRingOperation::Value operation) const
+{
+    return d_probe.isSupported(operation);
+}
+
+bool IoRingDevice::supportsCompletionQueueOverflow() const
+{
+    return d_params.supportsCompletionQueueOverflow();
+}
+
+bool IoRingDevice::supportsEnterTimeout() const
+{
+    return d_params.supportsEnterTimeout();
+}
+
+bool IoRingDevice::supportsNativeWorkers() const
+{
+    return d_params.supportsNativeWorkers();
+}
+
+bool IoRingDevice::supportsCancelByHandle() const
+{
+    return true;
```

Comment:
Why does it always return true?

Reply:
I wanted to wait until we merged standard support for the run-time detection of the operation system kernel version in `ntscfg_platform`. I have properly implemented detection of this type operation now.